sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -401,12 +399,41 @@ case class FileSourceScanExec(
       fsRelation: HadoopFsRelation): RDD[InternalRow] = {
     val defaultMaxSplitBytes =
       fsRelation.sparkSession.sessionState.conf.filesMaxPartitionBytes
-    val openCostInBytes = fsRelation.sparkSession.sessionState.conf.filesOpenCostInBytes
+    var openCostInBytes = fsRelation.sparkSession.sessionState.conf.filesOpenCostInBytes
     val defaultParallelism = fsRelation.sparkSession.sparkContext.defaultParallelism
     val totalBytes = selectedPartitions.flatMap(_.files.map(_.getLen + openCostInBytes)).sum
     val bytesPerCore = totalBytes / defaultParallelism
 
-    val maxSplitBytes = Math.min(defaultMaxSplitBytes, Math.max(openCostInBytes, bytesPerCore))
+    var maxSplitBytes = Math.min(defaultMaxSplitBytes, Math.max(openCostInBytes, bytesPerCore))
+    if(fsRelation.fileFormat.isInstanceOf[ParquetSource] &&
+      fsRelation.sparkSession.sessionState.conf.isParquetSizeAdaptiveEnabled) {
+      if (relation.dataSchema.map(_.dataType).forall(dataType =>
+        dataType.isInstanceOf[CalendarIntervalType] || dataType.isInstanceOf[StructType]
+          || dataType.isInstanceOf[MapType] || dataType.isInstanceOf[NullType]
+          || dataType.isInstanceOf[AtomicType] || dataType.isInstanceOf[ArrayType])) {
+
+        def getTypeLength (dataType : DataType) : Int = {
+          if (dataType.isInstanceOf[StructType]) {
+            fsRelation.sparkSession.sessionState.conf.parquetStructTypeLength
+          } else if (dataType.isInstanceOf[ArrayType]) {
+            fsRelation.sparkSession.sessionState.conf.parquetArrayTypeLength
+          } else if (dataType.isInstanceOf[MapType]) {
+            fsRelation.sparkSession.sessionState.conf.parquetMapTypeLength
+          } else {
+            dataType.defaultSize
+          }
+        }
+
+        val selectedColumnSize = requiredSchema.map(_.dataType).map(getTypeLength(_))
+          .reduceOption(_ + _).getOrElse(StringType.defaultSize)
+        val totalColumnSize = relation.dataSchema.map(_.dataType).map(getTypeLength(_))
+          .reduceOption(_ + _).getOrElse(StringType.defaultSize)
+        val multiplier = totalColumnSize / selectedColumnSize
```

Comment:
Seems here you can only get the ratio of selected columns to total columns. The actual type sizes are not put into consideration.

Reply:
There are many data types. CalendarIntervalType  StructType  MapType  NullType   UserDefinedType AtomicType(TimestampType  StringType  HiveStringType  BooleanType  DateType  BinaryType  NumericType)  ObjectType  ArrayType. For AtomicType, the size is fixed to the defaultSize. For complex type, such as StructType, MapType, ArrayType, the size is mutable. So I make it configurable with default value. With the data type size, multiplier is not only the ratio of selected columns to total columns, but the total size of selected columns to total size of all columns.

Reply:
@viirya  As defined in getTypeLength, user can define the complex types' length as per the data statistics. And the length for AtomicType can be determined by AtomicType.defaultSize. So the multiplier is the ratio of the total length of the selected columns to the total length of all columns.

def getTypeLength (dataType : DataType) : Int = {
          if (dataType.isInstanceOf[StructType]) {
            fsRelation.sparkSession.sessionState.conf.parquetStructTypeLength
          } else if (dataType.isInstanceOf[ArrayType]) {
            fsRelation.sparkSession.sessionState.conf.parquetArrayTypeLength
          } else if (dataType.isInstanceOf[MapType]) {
            fsRelation.sparkSession.sessionState.conf.parquetMapTypeLength
          } else {
            dataType.defaultSize
          }
        }

Reply:
@viirya  Now it also support ORC. Please help to review